Raft allows two leaders to be elected in the same term

On Jenkins we saw a failure of the should rejoin cluster test which shows in the log that two nodes became leader in the same term.

```
[ERROR] shouldRejoinCluster(io.zeebe.raft.RaftFiveNodesTest)  Time elapsed: 35.324 s  <<< FAILURE!
java.lang.AssertionError: Failed to wait for commit of event EventInfo{position=8589934816, term=2, message='world'} with message on all rafts
	at io.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:138)
	at io.zeebe.test.util.TestUtil$Invocation.until(TestUtil.java:86)
	at io.zeebe.test.util.TestUtil.waitUntil(TestUtil.java:58)
	at io.zeebe.raft.util.RaftClusterRule.awaitCondition(RaftClusterRule.java:277)
	at io.zeebe.raft.util.RaftClusterRule.awaitEventCommittedOnAll(RaftClusterRule.java:166)
	at io.zeebe.raft.RaftFiveNodesTest.shouldRejoinCluster(RaftFiveNodesTest.java:169)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)
```

[surefire-output.log](https://github.com/zeebe-io/zeebe/files/2213774/surefire-output.log)

--- a/qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/BrokerLeaderChangeTest.java
+++ b/qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/BrokerLeaderChangeTest.java
@@ -17,14 +17,18 @@ package io.zeebe.broker.it.clustering;
 
 import static io.zeebe.test.util.TestUtil.doRepeatedly;
 import static io.zeebe.test.util.TestUtil.waitUntil;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import io.zeebe.broker.it.ClientRule;
 import io.zeebe.gateway.api.commands.BrokerInfo;
+import io.zeebe.gateway.api.commands.PartitionBrokerRole;
+import io.zeebe.gateway.api.commands.PartitionInfo;
 import io.zeebe.gateway.api.events.JobEvent;
 import io.zeebe.gateway.api.events.JobState;
 import io.zeebe.gateway.api.subscription.JobWorker;
 import io.zeebe.gateway.api.subscription.TopicSubscription;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicBoolean;
 import org.junit.Ignore;
 import org.junit.Rule;
@@ -43,6 +47,33 @@ public class BrokerLeaderChangeTest {
   @Rule
   public RuleChain ruleChain =
       RuleChain.outerRule(testTimeout).around(clusteringRule).around(clientRule);
+
+  @Test
+  public void shouldBecomeFollowerAfterRestartLeaderChange() {
+    // given
+    final int partition = 1;
+    final int oldLeader = clusteringRule.getLeaderForPartition(partition).getNodeId();
+
+    clusteringRule.stopBroker(oldLeader);
+
+    waitUntil(() -> clusteringRule.getLeaderForPartition(partition).getNodeId() != oldLeader);
+
+    // when
+    clusteringRule.restartBroker(oldLeader);
+    clusteringRule.waitForTopic(3, 3);
+
+    // then
+    final Optional<PartitionInfo> partitionInfo =
+        clusteringRule
+            .getTopologyFromBroker(oldLeader)
+            .stream()
+            .filter(b -> b.getNodeId() == oldLeader)
+            .flatMap(b -> b.getPartitions().stream().filter(p -> p.getPartitionId() == partition))
+            .findFirst();
+
+    assertThat(partitionInfo)
+        .hasValueSatisfying(p -> assertThat(p.getRole()).isEqualTo(PartitionBrokerRole.FOLLOWER));
+  }
 
   @Test
   @Ignore("https://github.com/zeebe-io/zeebe/issues/844")
@@ -72,7 +103,7 @@ public class BrokerLeaderChangeTest {
     private final JobWorker jobSubscription;
     private final TopicSubscription topicSubscription;
 
-    JobCompleter(JobEvent jobEvent) {
+    JobCompleter(final JobEvent jobEvent) {
       final long eventKey = jobEvent.getMetadata().getKey();
 
       jobSubscription =

--- a/qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java
+++ b/qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java
@@ -521,16 +521,18 @@ public class ClusteringRule extends ExternalResource {
         () ->
             Arrays.stream(brokers)
                 .filter(Objects::nonNull)
-                .allMatch(
-                    b -> {
-                      final BrokerCfg config = b.getConfig();
-                      final List<BrokerInfo> topology =
-                          topologyClient.requestTopologyFromBroker(
-                              config.getNodeId(),
-                              config.getNetwork().getClient().toSocketAddress());
-                      return topologyPredicate.apply(topology);
-                    }),
+                .allMatch(b -> topologyPredicate.apply(requestTopology(b))),
         250);
+  }
+
+  public List<BrokerInfo> getTopologyFromBroker(final int nodeId) {
+    return requestTopology(brokers[nodeId]);
+  }
+
+  private List<BrokerInfo> requestTopology(Broker broker) {
+    final BrokerCfg config = broker.getConfig();
+    return topologyClient.requestTopologyFromBroker(
+        config.getNodeId(), config.getNetwork().getClient().toSocketAddress());
   }
 
   public SocketAddress getClientAddress() {

--- a/raft/src/main/java/io/zeebe/raft/Raft.java
+++ b/raft/src/main/java/io/zeebe/raft/Raft.java
@@ -115,7 +115,7 @@ public class Raft extends Actor
 
   private ServiceStartContext serviceContext;
 
-  private String raftName;
+  private final String raftName;
   private AbstractRaftState state;
   private RaftJoinService raftJoinedService;
 
@@ -143,7 +143,7 @@ public class Raft extends Actor
   }
 
   @Override
-  public void start(ServiceStartContext startContext) {
+  public void start(final ServiceStartContext startContext) {
     this.logStream = logStreamInjector.getValue();
     this.logStream.setTerm(getTerm());
 
@@ -156,7 +156,7 @@ public class Raft extends Actor
   }
 
   @Override
-  public void stop(ServiceStopContext stopContext) {
+  public void stop(final ServiceStopContext stopContext) {
     stopContext.async(actor.close());
   }
 
@@ -379,7 +379,7 @@ public class Raft extends Actor
         (l) -> LogUtil.catchAndLog(LOG, () -> l.onStateChange(this, getState())));
   }
 
-  private List<ActorFuture<Void>> notifyMemberLeavingListeners(List<Integer> newMembers) {
+  private List<ActorFuture<Void>> notifyMemberLeavingListeners(final List<Integer> newMembers) {
     final List<ActorFuture<Void>> futures = new ArrayList<>();
 
     raftStateListeners.forEach(
@@ -596,7 +596,7 @@ public class Raft extends Actor
           (t) -> {
             if (state.getState() == RaftState.LEADER) {
               raftMembers.removeMember(nodeId);
-              persistentStorage.save();
+
               // stop replication
               serviceContext.removeService(
                   replicateLogConrollerServiceName(raftName, getTerm(), nodeId));
@@ -635,7 +635,7 @@ public class Raft extends Actor
    *     at the moment
    */
   public ActorFuture<ClientResponse> sendRequest(
-      final int nodeId, final BufferWriter writer, Duration timeout) {
+      final int nodeId, final BufferWriter writer, final Duration timeout) {
     return clientTransport.getOutput().sendRequest(nodeId, writer, timeout);
   }
 

--- a/raft/src/main/java/io/zeebe/raft/RaftMembers.java
+++ b/raft/src/main/java/io/zeebe/raft/RaftMembers.java
@@ -29,7 +29,7 @@ public class RaftMembers {
   private final RaftPersistentStorage persistentStorage;
   private final RaftMember localMember;
 
-  public RaftMembers(int localNodeId, RaftPersistentStorage persistentStorage) {
+  public RaftMembers(final int localNodeId, final RaftPersistentStorage persistentStorage) {
     this.persistentStorage = persistentStorage;
     this.localMember = new RaftMember(localNodeId);
   }
@@ -46,11 +46,11 @@ public class RaftMembers {
     return members.size();
   }
 
-  public RaftMember getMember(int nodeId) {
+  public RaftMember getMember(final int nodeId) {
     return memberLookup.get(nodeId);
   }
 
-  public boolean hasMember(int nodeId) {
+  public boolean hasMember(final int nodeId) {
     return memberLookup.containsKey(nodeId);
   }
 
@@ -60,7 +60,7 @@ public class RaftMembers {
     memberLookup.clear();
     persistentStorage.clearMembers();
 
-    for (RaftConfigurationEventMember newMember : newMembers) {
+    for (final RaftConfigurationEventMember newMember : newMembers) {
       addMember(newMember.getNodeId());
     }
 
@@ -83,7 +83,7 @@ public class RaftMembers {
       members.add(member);
       memberLookup.put(nodeId, member);
 
-      persistentStorage.addMember(nodeId);
+      persistentStorage.addMember(nodeId).save();
 
       return member;
     } else {
@@ -100,7 +100,7 @@ public class RaftMembers {
     if (member != null) {
       members.remove(member);
       memberLookup.remove(nodeId, member);
-      persistentStorage.removeMember(nodeId);
+      persistentStorage.removeMember(nodeId).save();
     }
 
     return member;

